ExtensionException in OSGi environment for global extension

### Describe the bug

In an OSGi (e.g. Eclipse PDE) test execution spock throws an exception since change of #1995, which now tries to discover extensions from the ContextClassLoader and the Spock ClassLoader.

This now leads to the exception like:
```
Caused by: org.spockframework.runtime.extension.ExtensionException: Duplicated Extension declaration for [org.spockframework.runtime.extension.builtin.GlobalTimeoutExtension]
Source 1: jar:file:/<...>/spock-core-2.4-M5-groovy-4.0.jar!/META-INF/services/org.spockframework.runtime.extension.IGlobalExtension
Source 2: bundleresource://128.fwk1844173705/META-INF/services/org.spockframework.runtime.extension.IGlobalExtension
This is most likely caused by having two different version of a library on the classpath.
	at org.spockframework.runtime.ExtensionClassesLoader.loadClasses(ExtensionClassesLoader.java:46)
	at org.spockframework.runtime.ExtensionClassesLoader.loadExtensionClassesFromDefaultLocation(ExtensionClassesLoader.java:33)
	at org.spockframework.runtime.RunContext.createBottomContext(RunContext.java:230)
	at org.spockframework.runtime.RunContext.get(RunContext.java:195)
	at org.spockframework.runtime.SpockEngine.discover(SpockEngine.java:21)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:152)
```

This use due to the OSGi discrepancy of [MultiBundleClassLoader](https://github.com/eclipse-pde/eclipse.pde/blob/master/ui/org.eclipse.pde.junit.runtime/src/org/eclipse/pde/internal/junit/runtime/MultiBundleClassLoader.java) returning `jar:file:/` URLs, but the OSGi loader `bundleresource://` URLs. See change in [Eclipse PDE](https://github.com/eclipse-pde/eclipse.pde/commit/db0ba49ae4fc62da313310ce877ee0153da5301c#diff-d5f5b4aa3ce3c785fd85d5298c26e7d4bf09f4890772bdf7c9ae1dbe9a1bde3e), which uses `FileLocator.resolve()`.


We should probably change the behavior to the behavior of JUnit5 `org.junit.platform.launcher.core.ServiceLoaderRegistry.load()` 
which does only ask the `ContextClassLoader` and not the Spock ClassLoader, see `org.junit.platform.commons.util.ClassLoaderUtils.getDefaultClassLoader()`.

Currently the method `ReflectionUtil.getResourcesFromClassLoaders()` tries both, which lead to that issue.

So the proposal ist to change the method `getResourcesFromClassLoaders()` to only use the ContextClassLoader`, if present, otherwise use the Spock class loader, and do not merge both.





### To Reproduce

Try to load Spock 2.4-M5 in an Eclipse PDE/test env.

### Expected behavior

Spock shall be loadable.

### Actual behavior

org.spockframework.runtime.extension.ExtensionException: Duplicated Extension declaration for [org.spockframework.runtime.extension.builtin.GlobalTimeoutExtension]

### Java version

openjdk version "21.0.5" 2024-10-15 LTS
OpenJDK Runtime Environment Corretto-21.0.5.11.1 (build 21.0.5+11-LTS)
OpenJDK 64-Bit Server VM Corretto-21.0.5.11.1 (build 21.0.5+11-LTS, mixed mode, sharing)

### Buildtool version

Gradle 8.12

Build time:    2024-12-20 15:46:53 UTC
Revision:      a3cacb207fec727859be9354c1937da2e59004c1

Kotlin:        2.0.21
Groovy:        3.0.22
Ant:           Apache Ant(TM) version 1.10.15 compiled on August 25 2024
Launcher JVM:  21.0.5 (Amazon.com Inc. 21.0.5+11-LTS)
Daemon JVM:    D:\dev\.pes_eclipse\java\Amazon\21.0.5\win32\x86_64\jdk (no JDK specified, using current Java home)
OS:            Windows 11 10.0 amd64

### What operating system are you using

Windows

### Dependencies

None

### Additional context

_No response_

--- a/docs/release_notes.adoc
+++ b/docs/release_notes.adoc
@@ -5,6 +5,13 @@ include::include.adoc[]
 
 == 2.4 (tbd)
 
+== 2.4-M6 (tbd)
+
+=== Misc
+
+* Fix ExtensionException in OSGi environment for global extension spockIssue:2076[]
+** This issue was introduced with spockPull:1995[]
+
 == 2.4-M5 (2025-01-07)
 
 === Highlights

--- a/docs/release_notes.adoc
+++ b/docs/release_notes.adoc
@@ -3,8 +3,6 @@
 
 include::include.adoc[]
 
-== 2.4 (tbd)
-
 == 2.4-M6 (tbd)
 
 === Misc

--- a/spock-core/src/main/java/org/spockframework/runtime/ExtensionClassesLoader.java
+++ b/spock-core/src/main/java/org/spockframework/runtime/ExtensionClassesLoader.java
@@ -56,7 +56,7 @@ public class ExtensionClassesLoader {
 
   private Collection<URL> locateDescriptors(String descriptorPath) {
     try {
-      return ReflectionUtil.getResourcesFromClassLoaders(descriptorPath);
+      return ReflectionUtil.getResourcesFromClassLoader(descriptorPath);
     } catch (Exception e) {
       throw new ExtensionException("Failed to locate extension descriptors", e);
     }

--- a/spock-core/src/main/java/org/spockframework/util/ReflectionUtil.java
+++ b/spock-core/src/main/java/org/spockframework/util/ReflectionUtil.java
@@ -86,30 +86,24 @@ public abstract class ReflectionUtil {
   }
 
   /**
-   * Returns the resources from the following classloaders:
-   * <ul>
-   * <li>Spock {@link ClassLoader}</li>
-   * <li>{@code Thread.currentThread().getContextClassLoader()}</li>
-   * </ul>
+   * Returns the resources from the {@code Thread.currentThread().getContextClassLoader()}.
    *
    * @param resourcePath the path of the resource
    * @return the list of resources
    * @throws IOException if the resources can't be loaded
    */
-  public static Collection<URL> getResourcesFromClassLoaders(String resourcePath) throws IOException {
-    ClassLoader spockClassLoader = ReflectionUtil.class.getClassLoader();
-    // We need to use a sorted Set here, to filter out duplicates, if the ContextClassLoader can also reach the Spock classloader
-    TreeSet<URL> set = new TreeSet<>(Comparator.comparing(URL::toString));
-    set.addAll(Collections.list(spockClassLoader.getResources(resourcePath)));
+  public static Collection<URL> getResourcesFromClassLoader(String resourcePath) throws IOException {
     try {
-      //Also resolve resources via ContextClassLoader to better support for runtimes like OSGi
+      //Resolve resources via ContextClassLoader to better support for runtimes like OSGi
+      //Issue #2076: But du not use ContextClassLoader and the Spock ClassLoader and merge the results, because this fails for bundle resources
       ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
-      if (contextClassLoader != null && spockClassLoader != contextClassLoader) {
-        set.addAll(Collections.list(contextClassLoader.getResources(resourcePath)));
+      if (contextClassLoader != null) {
+        return Collections.list(contextClassLoader.getResources(resourcePath));
       }
     } catch (SecurityException ignored) {
     }
-    return set;
+    ClassLoader spockClassLoader = ReflectionUtil.class.getClassLoader();
+    return Collections.list(spockClassLoader.getResources(resourcePath));
   }
 
   public static boolean isClassAvailable(String className) {

--- a/spock-specs/src/test/groovy/org/spockframework/util/ReflectionUtilSpec.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/util/ReflectionUtilSpec.groovy
@@ -58,13 +58,14 @@ class ReflectionUtilSpec extends Specification {
     Thread.currentThread().setContextClassLoader(oldLoader)
   }
 
-  def "getResourcesFromClassLoaders"() {
+  def getResourcesFromClassLoader() {
     when:
-    def res = ReflectionUtil.getResourcesFromClassLoaders(JarFile.MANIFEST_NAME)
+    def res = ReflectionUtil.getResourcesFromClassLoader(JarFile.MANIFEST_NAME)
     then:
     res.size() >= 1
   }
 
+  @Issue("https://github.com/spockframework/spock/issues/2076")
   def "getResourcesFromClassLoaders - ContextClassLoader"() {
     given:
     def oldLoader = Thread.currentThread().getContextClassLoader()
@@ -74,26 +75,36 @@ class ReflectionUtilSpec extends Specification {
       @Override
       Enumeration<URL> getResources(String name) throws IOException {
         if (name == resPath) {
-          def url = new URL("file:/" + resPath)
-          return new Vector([
-            //Check that we filter out duplicates
-            url,
-            url
-          ]).elements()
+          return new Vector([new URL("file:/" + resPath)]).elements()
         }
         return super.getResources(name)
       }
     }
 
     expect:
-    ReflectionUtil.getResourcesFromClassLoaders(resPath).isEmpty()
+    ReflectionUtil.getResourcesFromClassLoader(resPath).isEmpty()
 
     when:
     Thread.currentThread().setContextClassLoader(resCl)
-    def res = ReflectionUtil.getResourcesFromClassLoaders(resPath)
+    def res = ReflectionUtil.getResourcesFromClassLoader(resPath)
     then:
     res.size() == 1
     res[0].toString().contains(resPath)
+
+    cleanup:
+    Thread.currentThread().setContextClassLoader(oldLoader)
+  }
+
+  @Issue("https://github.com/spockframework/spock/issues/2076")
+  def "getResourcesFromClassLoaders - ContextClassLoader == null"() {
+    given:
+    def oldLoader = Thread.currentThread().getContextClassLoader()
+
+    when:
+    Thread.currentThread().setContextClassLoader(null)
+    def res = ReflectionUtil.getResourcesFromClassLoader(JarFile.MANIFEST_NAME)
+    then:
+    res.size() >= 1
 
     cleanup:
     Thread.currentThread().setContextClassLoader(oldLoader)

